scipy/special/xsf/log_exp.h

```diff
@@ -27,7 +27,18 @@ inline float exprel(float x) { return exprel(static_cast<double>(x)); }
 
 template <typename T>
 T logit(T x) {
-    return std::log(x / (1 - x));
+    // The standard formula is log(x/(1 - x)), but this expression
+    // loses precision near x=0.5, as does log(x) - log1p(-x).
+    // We use the standard formula away from p=0.5, and use
+    // log1p(2*(x - 0.5)) - log1p(-2*(x - 0.5)) around p=0.5, which
+    // provides very good precision in this interval.
+    if (x < 0.3 || x > 0.65) {
```

Comment:
As $x \in \[0, 1\]$, should it be symmetric?
```suggestion
    if (x < 0.35 || x > 0.65) {
```
Setting values closer to 0.5 might improve performance, because most cases land in the first if condition (branch predictability) which only uses one logarithm.

Reply:
It sounds reasonable, since `logit` is symmetric around 0.5, but I think @WarrenWeckesser chose the cut offs empirically. To confirm, I've adapted his plot to put the relative error on a log scale, and to compare the direct formula and the improved formula for `x` near 0.5. You can see that there are still some spikes in relative error for the direct formula for $x \in (0.3, 0.35)$, but not for $x > 0.65$. I think the symmetry breaking is due to representable floating point numbers being more closely spaced in the interval $(0.3, 0.35)$ compared to in $(0.65, 0.7)$.

![logit_error](https://github.com/user-attachments/assets/39b741a2-c7e0-46b9-a047-1d367d42748a)

<details>

```python
import matplotlib.pyplot as plt
import numpy as np

from mpmath import mp


def reference_logit(x):
    x = mp.mpf(x)
    with mp.workdps(200):
        result = float(mp.log(x) - mp.log1p(-x))
    return result


def direct_formula(x):
    return np.log(x/(1 - x))


def new_formula(x):
    return np.log1p(2*(x - 0.5)) - np.log1p(-2*(x - 0.5))

p = np.linspace(0, 1, 8000)[1:-1]
y1 = direct_formula(p)
y2 = new_formula(p)
ref = np.array([reference_logit(t) for t in p])

relerr1 = abs(y1 - ref)/abs(ref)
relerr2 = abs(y2 - ref)/abs(ref)

plt.plot(p, np.log10(relerr1), label="direct")
plt.plot(p, np.log10(relerr2), label="new")
plt.legend()
plt.grid()
plt.xticks(np.linspace(0, 1, 21))
plt.xlabel('p')
plt.ylabel(r'$\log_{10}(\text{relerr})$')
plt.show()
```

</details>

Reply:
@steppi Thanks for the explanation.
I don't know scipy practice. In this case, I would have expected a short comment on the choice of the thresholds.

Reply:
We don't really have standard practices for the scalar kernels in `scipy.special` yet at that level of granularity since until recently it was a patchwork of different codebases (much of it rarely touched). But yes, with the ongoing efforts to standardize and the plan split it off into separate library that would be a common dependency of different array libraries, it's a good idea to start thinking of things like this. The work you've done with `wright_bessel` is some of the cleanest and best documented stuff we have in `special`, so I'd definitely be interested in hearing more of your thoughts on good practices. We've made a repo for the new library, https://github.com/scipy/xsf. I'm working on some RFCs that I'll post there in the coming weeks, and will make announcements on the scientific python discourse.

In any case, @mdhaber and I have a lot of trust that @WarrenWeckesser knows what he's doing, and from experience, I (and I'm sure also Matt) could tell right away how he must have chosen the cut offs between formulas. Rather than individual comments for each branch selecting a method within each special function, I wonder if it would be better to document generally how such cutoffs should be selected and to give links to the high precision reference implementations used for comparison.